src/wp-includes/blocks.php

```diff
@@ -797,34 +802,48 @@ function make_before_block_visitor( $context ) {
+			foreach ( $hooked_block_types as $hooked_block_type ) {
+				$hooked_block_markup = get_comment_delimited_block_content( $hooked_block_type, array(), '' );
+				/** This filter is documented in wp-includes/blocks.php */
+				$markup .= apply_filters( 'inject_hooked_block_markup', $hooked_block_markup, $hooked_block_type, $relative_position, $parent, $context );
 			}
 		}
 
-		$hooked_blocks = get_hooked_blocks( $block['blockName'] );
-		foreach ( $hooked_blocks as $hooked_block_type => $relative_position ) {
-			if ( 'before' === $relative_position ) {
-				$hooked_block_markup = get_comment_delimited_block_content( $hooked_block_type, array(), '' );
-				/**
-				 * Filters the serialized markup of a hooked block.
-				 *
-				 * @since 6.4.0
-				 *
-				 * @param string                  $hooked_block_markup The serialized markup of the hooked block.
-				 * @param string                  $hooked_block_type   The type of the hooked block.
-				 * @param string                  $relative_position   The relative position of the hooked block.
-				 *                                                     Can be one of 'before', 'after', 'first_child', or 'last_child'.
-				 * @param array                   $block               The anchor block.
-				 * @param WP_Block_Template|array $context             The block template, template part, or pattern that the anchor block belongs to.
-				 */
-				$markup .= apply_filters( 'inject_hooked_block_markup', $hooked_block_markup, $hooked_block_type, $relative_position, $block, $context );
-			}
+		$relative_position  = 'before';
+		$anchor_block_type  = $block['blockName'];
+		$hooked_block_types = array_keys( get_hooked_blocks( $anchor_block_type, $relative_position ) );
+		/** This filter is documented in wp-includes/blocks.php */
+		$hooked_block_types = apply_filters( 'hooked_block_types', $hooked_block_types, $relative_position, $anchor_block_type, $context );
+		foreach ( $hooked_block_types as $hooked_block_type ) {
+			$hooked_block_markup = get_comment_delimited_block_content( $hooked_block_type, array(), '' );
+			/**
+			 * Filters the serialized markup of a hooked block.
+			 *
+			 * @since 6.4.0
+			 *
+			 * @param string                  $hooked_block_markup The serialized markup of the hooked block.
+			 * @param string                  $hooked_block_type   The type of the hooked block.
+			 * @param string                  $relative_position   The relative position of the hooked block.
+			 *                                                     Can be one of 'before', 'after', 'first_child', or 'last_child'.
+			 * @param array                   $block               The anchor block.
+			 * @param WP_Block_Template|array $context             The block template, template part, or pattern that the anchor block belongs to.
+			 */
+			$markup .= apply_filters( 'inject_hooked_block_markup', $hooked_block_markup, $hooked_block_type, $relative_position, $block, $context );
```

Comment:
@gziolo We might consider removing the `inject_hooked_block_markup` filter in a follow-up; it might be a bit low-level (and less useful?) after all.

Reply:
`inject_hooked_block_markup` operates on the markup level, so it basically allows setting any string as the return value. The difference in the params available in the filter is that `hooked_block_types` uses the block name as the anchor, in contrast to the block type object that `inject_hooked_block_markup` offers for the parent or sibling block. I would be in favor of keeping only `hooked_block_types` which is a nice abstraction as it's just a list of block names.